### DISCUSSION TITLE
Saving a tiddler from JS

### DISCUSSION THREAD

**Discussion by EvidentlyCube** [TiddlyWiki/TiddlyWiki5#6093] [Q&A] [answered]
I am writing a plugin which will propagate some information about a field upwards. Specifically, I want to add to certain tiddlers a field `private` and all the "parent" tiddlers will have a field "has-private-child" which will be dynamically updated on save. 

Going up and down the *tree of tags* is not a problem, I've got the code done. My problem is how to trigger the tiddler save on all the updated parents? Normally save seems to be triggered by an action sending a message which is captured by NavigatorWidget, which consumes the message and does some things.

On the other hand, maybe all I have to do is call `$tw.wiki.addTiddler(newTiddler);`, because I know the name won't be changed and I know there will be no name collision?

**Comment by saqimtiaz** [comment, marked as answer]
Updating and saving a new tiddler is akin to saving a new tiddler with the same title.
```
var oldTiddler= $tw.wiki.getTiddler(title);
var newFields = {
    parent: parentTitle
}
$tw.wiki.addTiddler(new $tw.Tiddler(oldTiddler,newFields)); //right most hashmaps take precedence

```

**Comment by pmario** [comment]
>I am writing a plugin which will propagate some information about a field upwards. Specifically, I want to add to certain tiddlers a field private and all the "parent" tiddlers will have a field "has-private-child" which will be dynamically updated on save.

hmmmm, I personally would try to create a different structure or store the "has-private-child" info differently. You will get all sorts of synchronisation problems. Especially if tiddlers are renamed or the child tiddler is moved or deleted and so on. ... You are free to ignore this comment. .. Just think back when the problems start.

**Reply by EvidentlyCube** [reply]
> Just think back **when** the problems start.

I see that pesky "when" there :).

I have thought about those issues (but I missed the delete one!) I actually wanted to do a full-tree update after the save is finished but there is no hook for that. But now that I think about it, since auto save event is always dispatched after deleting or saving a tiddler, I could catch that before `SaverHandler` does and then do the full refresh.

Let's see if I can make it work.

Thank you!

**Reply by saqimtiaz** [reply]
I have started using sidecar tiddlers for storing relational data between tiddlers. So instead of saving the data in the parent or the children, you store it in a separate tiddler. Handy when you don't want changing the relationships between tiddlers to modify the tiddlers in question.

**Reply by pmario** [reply]
OK. I don't know your data structure, but there is a `th-deleting-tiddler` hook, that you can catch. ... The navigator also has a tm-rename-tiddler message, which directly calls `$tw.wiki.renameTiddler()` which does a delete and recreate ... BUT without the possibility for you to catch it. ...

**Reply by pmario** [reply]
Sorry. There is a hook: `th-renaming-tiddler`

**Reply by EvidentlyCube** [reply]
Thanks, but I managed to do it with `tm-auto-save-wiki` event. Since it runs after all the changes have been propagated to the internal state there's no worry anything will be wrong. The full refresh is a bit less optimal but I am not really worried about it.
